relentless/simulate/generic.py

```diff
@@ -0,0 +1,296 @@
+import importlib
+import inspect
+
+import numpy as np
+
+from . import simulate
+
+class SimulationOperationAdapter(simulate.SimulationOperation):
+    """Generic simulation operation adapter.
+
+    Generic simulation operation is used to call simulation operation for a backend
+    class of type :py:class:`Simulation`.
+
+    Parameters
+    ----------
+    args : args
+        Positional arguments for simulation operation.
+    kwargs : kwargs
+        Keyword arguments for simulation operation.
+
+    """
+    backends = {}
+
+    def __init__(self, *args, **kwargs):
+        self.args = args
+        self.kwargs = kwargs
+        self._op = None
+        self._backend = None
+
+    def __call__(self, sim):
+        if self._op is None or self._backend != sim.backend:
+            backend = self.backends.get(sim.backend)
+            if not backend:
+                raise TypeError('Simulation backend {} not registered.'.format(backend))
+
+            op_name = type(self).__name__
+            try:
+                BackendOp = getattr(backend,op_name)
+            except AttributeError:
+                raise TypeError('{}.{}.{} operation not found.'.format(backend.__module__,backend.__name__,op_name))
+
+            self._op = BackendOp(*self.args,**self.kwargs)
+            self._backend = sim.backend
+
+        return self._op(sim)
+
+    @classmethod
+    def add_backend(cls, backend, module=None):
+        """Adds backend attribute to :py:class:`SimulationOperationAdapter`.
+
+        Parameters
+        ----------
+        backend : :py:class:`Simulation`
+            Class to add as a backend.
+        module : ??
+            Name of the module in which the backend is defined (defaults to `None`).
+
+        """
+        # try to deduce module from backend class
+        if module is None:
+            module = backend.__module__
+
+        # setup module (if not already a module)
+        if not inspect.ismodule(module):
+            module = importlib.import_module(module)
+
+        cls.backends[backend] = module
+
+## initializers
+class InitializeFromFile(SimulationOperationAdapter):
+    """Initializes a simulation box and pair potentials from a file.
+
+    Parameters
+    ----------
+    filename : str
+        The file from which to read the system data.
+    options : kwargs
+        Options for file reading.
+    neighbor_buffer : float
+        Buffer width for neighbor list (defaults to 0.4).
```

Comment:
Please reorder these. Also, let's remove the default value of `neighbor_buffer`. The user should be explicit about this choice. You should also remove this default in the `hoomd` module.